Mathlib/LinearAlgebra/Matrix/SVD/Defs.lean

```diff
@@ -0,0 +1,479 @@
+
+This file provides the SVD theorem which decomposes a real/complex matrix into left
+eigenvectors, singular values block diagonal matrix and right eigenvectors.
+
+Any matrix A (M × N) with rank r = A.rank and with elements in the field ℝ or ℂ can be decomposed
+into three matrices:
+  U: an M × M matrix containing the left eigenvectors of the matrix
+  S: an M × N matrix with an r × r block in the upper left corner with nonzero singular values
+  V: an N × N matrix containing the right eigenvectors of the matrix
+Note that
+  S is a block matrix S = [S₁₁, S₁₂; S₂₁, S₂₂] with
+  - S₁₁: a diagonal r × r matrix and
+  - S₁₂: r × (N - r) zero matrix, S₂₁ : (M - r) × r zero matrix and
+  - S₂₂: (M - r) × (N - r) zero matrix
+  U is a block column matrix U = [U₁ U₂] with
+  - U₁ : a M × r containing left eigenvectors with nonzero singular values.
+  - U₂ : a M × (M - r) containing left eigenvectors with zero singular values.
+  V is a block column matrix V = [V₁ V₂] with
+  - V₁ : a N × r containing right eigenvectors with nonzero singular values.
+  - V₂ : a M × (M - r) containing right eigenvectors with zero singular values.
+
+Since in mathlib the eigenvalues of hermitian matrices are defined in an "arbitrary" undetermined
+order, we begin by partitioning the singular values into zero and non-zero values. We partition the
+corresponding eigenvectors from AᴴA and AAᴴ using similar rearrangements. These are included in
+`SVD.svdReindex`. The basic API for Column and Row partitioned matrices is from
+`ColumnRowPartitioned`.
+
+We then proceed to transfer some of the lemmas we need about eigenvector matrices (for example that
+they are unitary: i.e. inverse is conjugate transpose.). Note that since invertibility in mathlib is
+defined for square matrices while our matrices are partitioned i.e. N × (N₁ ⊕ N₂) and N ≃ (N ⊕ N₂)
+Lean cannot apply the Invertible definition. We work around this where necessary.
+
+Lemma `reduced_spectral_theorem` (`reduced_spectral_theorem'`) shows that AᴴA and AAᴴ can be
+reduced to products containing only the non-zero singular eigenvectors. This is later used in
+proving the main SVD theorem. A few lemmas are provided about the invertibility of the non-zero
+singular values matrix: `svdσ_inv`, `σ_inv_μ_σ_inv_eq_one`, `IsUnit_det_svdσ`,
+`IsUnit_det_svdσ_mapK` and `svdσ_inv_mapK`.
+
+To make relating left eigenvectors to right eigenvectors easier we define U₁ = AV₁σ⁻¹ while U₂ is
+obtained from the eigenvectors of (AAᴴ). This avoid a lengthy reindexing operation with many proofs.
+The vectors in U₂ have no such issue since they are multiplied by zero singular values anyway.
+
+## Tags
+Singular Value decomposition, SVD
+-/
+
+
+variable {𝕂 : Type*} [IsROrC 𝕂] [DecidableEq 𝕂]
+variable {M N : ℕ}
+
+open Matrix BigOperators
+
+namespace Matrix
+
+/-- The right eigenvectors of a matrix A corresponding to its non-zero eigenvaules -/
+noncomputable def svdV₁ (A : Matrix (Fin M) (Fin N) 𝕂) : Matrix (Fin N) (Fin (A.rank)) 𝕂 :=
```

Comment:
I prefer that the decomposed matrices be usable through the matrix. For example I want `A.svdV₁` to work. When I tried your suggestion of using a namespace I was not able to get this working. I had to use (Matrix.SVD.V₁ A). 

For example if you try this you will get an error 

```lean

variable {𝕂 : Type*} [IsROrC 𝕂] [DecidableEq 𝕂]
variable {M N : ℕ}

open Matrix BigOperators

namespace Matrix
namespace SVD
variable (A : Matrix (Fin M) (Fin N) 𝕂)

noncomputable def V₁ :  Matrix (Fin N) (Fin (A.rank)) 𝕂 :=
  ((reindex (Equiv.refl (Fin N)) (eigenColumnEquiv A))
    (isHermitian_transpose_mul_self A).eigenvectorMatrix).toColumns₁

noncomputable def U₁ := A.SVD.V₁

end SVD
end Matrix

```

The error says 
```
invalid field 'SVD', the environment does not contain 'Matrix.SVD'
  A
has type
  Matrix (Fin M) (Fin N) 𝕂
```

Reply:
That's indeed slightly annoying. However, you can still `open Matrix` and then `SVD.V₁ A` and `A.SVD.V₁` don't really make any difference. Here I would really advocate the consistency with `LDL` and the neater organisation over the notational advantages (especially since we're already quite far away from standard maths notation). So I'm going to insist on the `SVD` namespace.